crypto/merklearray/proof.go

```diff
@@ -41,12 +42,33 @@ type Proof struct {
 // SingleLeafProof is used to convince a verifier about membership of a specific
 // leaf h at index i on a tree. The verifier has a trusted value of the tree
 // root hash. it corresponds to merkle verification path.
+//
+//msgp:maxsize ignore SingleLeafProof
 type SingleLeafProof struct {
 	_struct struct{} `codec:",omitempty,omitemptyarray"`
 
 	Proof
 }
 
+// ProofMaxSizeByElements returns the maximum msgp encoded size of merklearray.Proof structs containing n signatures
+// This is necessary because the allocbounds on the proof are actual theoretical bounds but for calculating maximum
+// proof size for individual message types we have smaller valid bounds.
+// Exported because it's used in the stateproof package for ensuring that SigPartProof constant is correct size.
+func ProofMaxSizeByElements(n int) (s int) {
+	s = 1 + 4
```

Comment:
It's actually not fully manual. It's a modification of the autogenerated test: here https://github.com/algorand/go-algorand/pull/5388/files#diff-add6a5ef46c2629eb7484bf3113df670d70cfdb6661ca69e236a778a7e1cb4c4R334

and we have a test here https://github.com/algorand/go-algorand/pull/5388/files#diff-181a1af1a38a976d2ddce9a94f778520e2ad787079c3cc99dbd1b7483386aa73R248 that ensures that that auto-generated ProofMaxSize and and ProofMaxSizeByElements output the same value when the allocbound is used as the input

Reply:
The test is very good and reliable. However, the mapping of the function `SingleLeafProofMaxSize` to `SingleLeafProof` is not obvious for someone who is not aware of this arrangement. 

In this particular case, this function is not needed. It can be implemented by setting a constant bound directly to the Proof field in:
```
        Signature struct {
                _struct struct{} `codec:",omitempty,omitemptyarray"`

                Signature             crypto.FalconSignature      `codec:"sig"`
                VectorCommitmentIndex uint64                      `codec:"idx"`
                Proof                 merklearray.SingleLeafProof `codec:"prf"`
```

Reply:
Yeah but we do have a pattern for doing this manual implementation already for basics.MicroAlgos and this solution will continue working if we ever use SingleLeafProof elsewhere. Finding it's `maxtotalbytes` bound and using it is more implicit IMO.